expan/core/binning.py

```diff
@@ -8,725 +6,243 @@
-        the latest bins that match will take precedence.
-        """
-        data = np.array(data)
-
-        try:
-            out = np.empty(data.shape, int)
-        except AttributeError:  # duck typing
-            data = np.array(data)
-            out = np.empty(data.shape, int)
-
-        logger.debug('apply: _uppers: ' + str(self._uppers))
-        logger.debug('apply: _lowers: ' + str(self._lowers))
-        logger.debug('apply: _mids: ' + str(self._mids))
-        logger.debug('apply: _up_closed: ' + str(self._up_closed))
-        logger.debug('apply: _lo_closed: ' + str(self._lo_closed))
-
-        out[:] = -1
-        for ii, (lc, uc, l, u) in enumerate(
-            zip(self.lo_closed, self.up_closed, self.lowers, self.uppers)):
-            if np.isnan(l) or np.isnan(u):  # if either bound is nan, only nans exist in the bin.
-                got = np.isnan(data)
-            else:
-                got = (l <= data) if lc else (l < data)
-                got &= (data <= u) if uc else (data < u)
-
-            logger.debug('apply: testing {:d} {}: {:d} members'.format(ii, self._labels('{conditions}')[ii], got.sum()))
-            out[got] = ii
-
-        return out
-
-    def _bound(self, data, which):
-        if which == 'upper':
-            bounds = self._uppers
-        elif which == 'lower':
-            bounds = self._lowers
-        elif which == 'mid':
-            bounds = self._mids
+
+#------- bin representations -------#
+class Bin(object):
+    # this is a necessary hack for the buggy implementation of assertItemsEqual in python2
+    # see https://stackoverflow.com/a/29690198
+    # note that if we only use python3, assertCountEqual in python3 solves this problem
+    __hash__ = None
+
+    def __init__(self, id, bin_type, representation):
+        '''
+        Constructor for a bin object.
+        :param id: identifier (e.g. bin number) of the bin
+        :param bin_type: "numerical" or "categorical"
+        :param representation: representation of the bin. can be NumericalRepresenation or CategoricalRepresenation.
+        '''
+        if bin_type == "numerical" and type(representation) is not NumericalRepresenation:
+            raise ValueError("Use NumericalRepresenation object to initialize a numerical bin.")
+        if bin_type == "categorical" and type(representation) is not CategoricalRepresenation:
+            raise ValueError("Use CategoricalRepresenation object to initialize a categorical bin.")
+        self.id = id
+        self.bin_type = bin_type
+        self.representation = representation
+
+    def __repr__(self):
+        return "\nbin id: " + str(self.id) + ", bin: " + str(self.representation)
+
+    def __eq__(self, other):
+        return isinstance(other, self.__class__) and self.__dict__ == other.__dict__
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+
+class NumericalRepresenation(object):
```

Comment:
smaller typo in the Numerical**Represenation** and Numerical**Represenation**.

Reply:
thanks